security/rbac/service.go

```diff
@@ -0,0 +1,157 @@
+package rbac
+
+import (
+	"errors"
+	"log"
+
+	"github.com/casbin/casbin/v2"
+	"github.com/casbin/casbin/v2/model"
+	"github.com/casbin/casbin/v2/persist"
+	"gopkg.in/yaml.v2"
+
+	"github.com/gaia-pipeline/gaia"
+	"github.com/gaia-pipeline/gaia/helper/assethelper"
+)
+
+type (
+	// RoleRule represents a Casbin role rule line in the format we expect.
+	RoleRule struct {
+		Namespace string `json:"namespace"`
+		Action    string `json:"action"`
+		Resource  string `json:"resource"`
+		Effect    string `json:"effect"`
+	}
+
+	// Service wraps the Casbin enforcer and performs all actions we require to manage and use RBAC functions.
+	Service interface {
+		EndpointEnforcer
+		AddRole(role string, roleRules []RoleRule) error
+		DeleteRole(role string) error
+		GetAllRoles() []string
+		GetUserAttachedRoles(username string) ([]string, error)
+		GetRoleAttachedUsers(role string) ([]string, error)
+		AttachRole(username string, role string) error
+		DetachRole(username string, role string) error
+	}
+
+	enforcerService struct {
+		adapter         persist.BatchAdapter
+		enforcer        casbin.IEnforcer
+		rbacapiMappings gaia.RBACAPIMappings
+	}
+)
+
+// NewEnforcerSvc creates a new EnforcerService.
+func NewEnforcerSvc(adapter persist.BatchAdapter) (Service, error) {
+	model, err := loadModel()
+	if err != nil {
+		return nil, err
+	}
+
+	enforcer, err := casbin.NewEnforcer(model, adapter)
+	if err != nil {
+		log.Fatal(err)
+	}
+	enforcer.EnableLog(true)
+
+	rbacapiMappings, err := loadAPIMappings()
+	if err != nil {
+		return nil, err
+	}
+
+	return &enforcerService{
+		enforcer:        enforcer,
+		rbacapiMappings: rbacapiMappings,
+	}, nil
+}
+
+func loadModel() (model.Model, error) {
+	modelStr, err := assethelper.LoadRBACModel()
+	if err != nil {
+		return nil, err
+	}
+
+	model, err := model.NewModelFromString(modelStr)
+	if err != nil {
+		return nil, err
+	}
+
+	return model, nil
+}
+
+func loadAPIMappings() (gaia.RBACAPIMappings, error) {
+	mappings, err := assethelper.LoadRBACAPIMappings()
+	if err != nil {
+		return gaia.RBACAPIMappings{}, err
+	}
+
+	var rbacapiMappings gaia.RBACAPIMappings
+	if err := yaml.Unmarshal([]byte(mappings), &rbacapiMappings); err != nil {
```

Comment:
I wonder if we should lock this mapping down for tempering with UnmarshalStrict. That will throw errors in case of duplicate fields and mappings which do not have corresponding structs.